libraries/libexosphere/source/fuse/fuse_api.cpp

```diff
@@ -0,0 +1,426 @@
+        }
+    }
+
+    PatchVersion GetPatchVersion() {
+        const auto patch_version = reg::Read(GetChipRegisters().FUSE_SOC_SPEEDO_1_CALIB);
+        return static_cast<PatchVersion>(static_cast<int>(GetSocType() << 12) | patch_version);
+    }
+
+    QuestState GetQuestState() {
+        return static_cast<QuestState>(util::BitPack32{GetOdmWord(4)}.Get<OdmWord4::QuestState>());
+    }
+
+    pmic::Regulator GetRegulator() {
+        /* TODO: How should mariko be handled? This reads from ODM word 28 in fuses (not present in erista...). */
+        return pmic::Regulator_Erista_Max77621;
+    }
+
+    int GetDeviceUniqueKeyGeneration() {
+        if (IsNewFuseFormat()) {
+            return util::BitPack32{GetOdmWord(2)}.Get<OdmWord2::DeviceUniqueKeyGeneration>();
+        } else {
+            return 0;
+        }
+    }
+
+    SocType GetSocType() {
+        switch (GetHardwareType()) {
+            case HardwareType_Icosa:
+            case HardwareType_Copper:
+                return SocType_Erista;
+            case HardwareType_Iowa:
+            case HardwareType_Hoag:
+            case HardwareType_Calcio:
+            case HardwareType_Five:
+                return SocType_Mariko;
+            default:
+                return SocType_Undefined;
+        }
+    }
+
+    int GetExpectedFuseVersion(TargetFirmware target_fw) {
+        return GetExpectedFuseVersionImpl(target_fw);
+    }
+
+    bool HasRcmVulnerabilityPatch() {
+        /* Only check for RCM bug patch once, and cache our result. */
+        if (!g_checked_for_rcm_bug_patch) {
+            do {
+                /* Mariko units are necessarily patched. */
+                if (fuse::GetSocType() != SocType_Erista) {
+                    g_has_rcm_bug_patch = true;
+                    break;
+                }
+
+                /* Some patched units use XUSB in RCM. */
+                if (reg::Read(GetChipRegisters().FUSE_RESERVED_SW) & 0x80) {
+                    g_has_rcm_bug_patch = true;
+                    break;
+                }
+
+                /* Other units have a proper ipatch instead. */
+                u32 word_count = reg::Read(GetChipRegisters().FUSE_FIRST_BOOTROM_PATCH_SIZE) & 0x7F;
+                u32 word_addr  = 191;
+
+                while (word_count && !g_has_rcm_bug_patch) {
+                    u32 word0        = ReadWord(word_addr);
+                    u32 ipatch_count = (word0 >> 16) & 0xF;
+
+                    for (u32 i = 0; i < ipatch_count && !g_has_rcm_bug_patch; ++i) {
+                        u32 word = ReadWord(word_addr - (i + 1));
+                        u32 addr = (word >> 16) * 2;
+
+                        if (addr == 0x769a) {
+                            g_has_rcm_bug_patch = true;
+                            break;
+                        }
+                    }
+
+                    word_addr -= word_count;
+                    word_count = word0 >> 25;
+                }
+            } while (0);
```

Comment:
Why the while(0)?

Reply:
This allows the break statements to work to early-exit the logic.